topdown/eval.go

```diff
@@ -3089,3 +3142,20 @@ func refContainsNonScalar(ref ast.Ref) bool {
 	}
 	return false
 }
+
+type signalEarlyExit struct {
+	ref ast.Ref
+}
+
+func (e signalEarlyExit) Error() string {
+	return fmt.Sprintf("%v: early exit", e.ref)
+}
+
+func isEarlyExit(err error, ref ast.Ref) bool {
```

Comment:
I don't understand why we need the ref to determine if the early exit signal/error is from the current evaluation context... is there a test case that asserts on this behaviour? I tried disabling the check below (so that it just returns `ok`) and all of the topdown tests passed.

Since we're returning the early-exit error inside of the iterator for the child query and we're catching the error and returning nil inside of the caller that started the child query, I don't see how the early-exit error could leak out (and then be caught by the parent of the caller... if you see what I mean...)

Reply:
> I don't see how the early-exit error could leak out (and then be caught by the parent of the caller... if you see what I mean...)

I'll double check, but if I remember correctly the problem was having a child of the caller react to the error. If one evalVirtualComplete's eval required another evalVirtualComplete, and finished successfully, the inner one shouldn't necessarily exit early.

But there might be other ways... Perhaps we need to check if that inner one has e.ir.ExitEarly ... Need to play more with this.

Reply:
I've check: in the caller tree, this is used where the "EE signal" is consumed: an error was returned, but we decide that it's OK, and go along.

In this example:

```rego
package test
p { q; trace("a") }
p { q; trace("b") }
q { trace("c") }
q { trace("d") }`,
```

the call tree is roughly
- evalVirtualComplete for `p`
  - evalVirtualComplete for `q`
    - the `iter` closure for `p` return err=EE
  - it's checked here, and passed along because it's not the one belonging to `q`
- it's checked here, and it's accepted, returning `nil` instead

Without comparing the refs, the first check would "discharge" the error and we'd see a trace note "b" for the second `p` body.

Reply:
Makes sense. I've got some FUD (fear-uncertainty-doubt) about using the refs to track this--but I can't point to anything specifically wrong in the implementation. Thanks for looking into it and improving the test coverage.

Reply:
Never wrong to brainstorm alternatives. Basically, we'll need a way to _throw_ something in some inner-most iterator, and _catch_ it in that evalVirtualComplete's iteration call.

Right now, we're returning an error marked with the ref. 

I think another option would be to use the eval ptr value. I'll explore that. At any rate, it should be completely unambiguous.

Reply:
PTAL, I've made it a struct field instead. There should be no room for error now when it comes it who's got to deal with the errEarlyExit.

Reply:
I like it! Can we do this without struct embedding though? We use struct embedding sparingly in the rest of the code base (it's often unnecessary and just makes the Go harder to read IMO.)

Reply:
Yeah sure. I'll push an update later.